nixos/modules/services/security/opensnitch.nix

```diff
@@ -0,0 +1,174 @@
+          data = "${pkg}/.*";
+    };
+
+  makeHostRuleFile = rulePath: type: value: let
+    name = "nixos-allow-${type}-${value}";
+    in makeAlwaysOperatorFile rulePath name {
+      type = "regexp";
+      operand = "dest.${type}";
+      data = replaceStrings [ "." "*" ] [ "\\." ".*" ] value;
+    };
+
+in
+
+{
+
+  options = {
+    services.opensnitch = with lib; {
+      enable = mkEnableOption "opensnitch";
+
+      uiConfig = mkOption {
+        type = types.attrs;
+        description = "JSON attribute set for opensnitch-ui config file.";
+        default = {
+          default_timeout = 60;
+          default_action = "deny";
+          default_duration = "once";
+        };
+        example = ''
+          {
+            default_timeout = 15;
+            default_action = "allow";
+            default_duration = "until restart";
+          }
+        '';
+      };
+
+      startUserService = mkOption {
+        type = types.bool;
+        description = ''
+          If enabled, run the opensnitch-ui process as a user
+          service in the graphical session.  All users' UI processes
+          share the config specified in uiConfig.  If this is disabled,
+          opensnitch-ui must be run by some other means.
+        '';
+        default = true;
+      };
+
+      whitelistPackages = mkOption {
+        type = types.listOf types.package;
+        description = ''
+          List of packages for which to generate rules to allow connections
+          from all processes that are located below
+          a package's store path.  Intended for process rules
+          which should survive NixOS updates.
+        '';
+        default = with pkgs; [ nix ];
+      };
+
+      whitelistHosts = mkOption {
+        type = types.listOf types.attrs;
+        description = ''
+          List of destination hosts for which to create default (regexp) allow rules,
+          regardless of other connection properties.
+        '';
+        default = [];
+        example = ''[ { host = "*.nixos.org" } { ip = "127.0.0.1" } ]'';
+      };
+
+      extraRules = mkOption {
+        type = types.attrsOf types.attrs;
+        description = ''
+          Set of JSON attribute sets describing default opensnitch rules that are written to
+          /etc/opensnitch/rules.
+        '';
+        default = [];
+        example = ''
+          {
+            kerberos_rule = {
+              action = "allow";
+              operator = {
+                type = "list";
+                list = [
+                  {
+                    type = "simple";
+                    operand = "dest.host";
+                    data = "kerberos.example.domain";
+                  }
+                  {
+                    type = "simple";
+                    operand = "dest.port";
+                    data = "88";
+                  }
+                ];
+              };
+            };
+          }
+        '';
+      };
+    };
+  };
+
+  config =
```

Comment:
Should we `assert` that the NixOS firewall is off?

Reply:
You mean that the firewall is enabled? As far as I know both ``iptables`` and ``nftables`` are supported by Opensnitch

Reply:
Isn't this program meant to manage your system firewall? Won't nixos firewall rules collide?